golem-worker-service-base/src/gateway_api_definition/http/rib_converter.rs

```diff
@@ -0,0 +1,249 @@
+                        if let Some(case_schema) = self.convert_type(typ) {
+                            let mut case_obj = Object::with_type(Type::Object);
+                            case_obj.properties = BTreeMap::new();
+                            case_obj.properties.insert(case.name.clone(), RefOr::T(case_schema));
+                            schemas.push(Schema::Object(case_obj));
+                        }
+                    }
+                }
+
+                if !schemas.is_empty() {
+                    let mut obj = Object::with_type(Type::Object);
+                    obj.description = Some("Variant type".to_string());
+                    obj.properties = BTreeMap::new();
+
+                    let discriminator_obj = Object::with_type(Type::String);
+                    obj.properties.insert("discriminator".to_string(), RefOr::T(Schema::Object(discriminator_obj)));
+
+                    let mut one_of = OneOf::new();
+                    for schema in schemas {
+                        one_of.items.push(RefOr::T(schema));
+                    }
+                    obj.properties.insert("value".to_string(), RefOr::T(Schema::OneOf(one_of)));
+                    Some(Schema::Object(obj))
+                } else {
+                    None
+                }
+            }
+            AnalysedType::Option(option_type) => {
+                if let Some(inner_schema) = self.convert_type(&option_type.inner) {
+                    let mut obj = Object::with_type(Type::Object);
+                    obj.description = Some("Optional value".to_string());
+                    obj.properties = BTreeMap::new();
+                    obj.properties.insert("value".to_string(), RefOr::T(inner_schema));
+                    obj.required = vec![];
+                    Some(Schema::Object(obj))
+                } else {
+                    None
+                }
+            }
+            AnalysedType::Result(result_type) => {
+                let mut properties = BTreeMap::new();
+
+                if let Some(ok_type) = &result_type.ok {
+                    if let Some(ok_schema) = self.convert_type(ok_type) {
+                        properties.insert("ok".to_string(), RefOr::T(ok_schema));
+                    }
+                }
+
+                if let Some(err_type) = &result_type.err {
+                    if let Some(err_schema) = self.convert_type(err_type) {
+                        properties.insert("err".to_string(), RefOr::T(err_schema));
+                    }
+                }
+
+                if !properties.is_empty() {
+                    let mut obj = Object::with_type(Type::Object);
+                    obj.properties = properties;
+                    obj.description = Some("Result type".to_string());
+                    Some(Schema::Object(obj))
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        }
+    }
+}
+
+#[cfg(test)]
```

Comment:
This test suite is not sufficient to prove the conversion is correct.

I would exhaustively enumerate every possible WIT type, and for each, verify the conversion is correct.

Then I would add some more complex, hand-constructed composite types--for example, variants containing records, containing lists, containing variants containing records, containing flags and primitive types. Just a couple to verify the machinery works there.

Finally, there's the question of "verify": how do you verify the WIT type has the JSON type being described by the OpenAPI Schema generated? Obviously, you have to rely on the "canonical JSON representation of WIT values", which is baked into the Worker Gateway and which it uses to render WIT values to JSON.

But how to test that the schema of the converted WIT value is in fact the same as the inferred schema here?

I would suggest:

1. Find the code which converts Rib values to JSON (don't recall where it is)
2. For each type, construct a Rib script that has the type (just a simple Rib value, e.g. '1: u16') for an integer
3. Compile / evaluate the Rib script to get back a Rib value
4. Convert the Rib value to JSON
5. Verify the JSON value conforms to the schema

This is a robust way to test that your knowledge of the canonical JSON mapping is correct. Robust because as the code changes, regressions will be detected.